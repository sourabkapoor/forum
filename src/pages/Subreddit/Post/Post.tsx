import styles from "./Post.module.css";
import { BiUpvote } from "react-icons/bi";
import { BiDownvote } from "react-icons/bi";
import { BiMessage } from "react-icons/bi";
import Votes from "../Votes/Votes";
import PostAuthor from "../PostAuthor/PostAuthor";
import PostInteractions from "../PostInteractions/PostInteractions";
import {
  addDoc,
  collection,
  doc,
  DocumentData,
  getDoc,
  getDocs,
  onSnapshot,
  query,
  QuerySnapshot,
  runTransaction,
  serverTimestamp,
  setDoc,
  where,
  writeBatch,
} from "firebase/firestore";
import CSSModules from "react-css-modules";
import { render } from "@testing-library/react";
import { db, getUserId } from "../../../firebase";
import React, { useEffect, useReducer, useState } from "react";
import { useLocation, useParams } from "react-router-dom";
import { useDispatch } from "react-redux";
import {
  selectAuthStatus,
  toggleSignInModal,
} from "../../../features/auth/authSlice";
import { nanoid } from "nanoid";
import { useAppSelector } from "../../../hooks/hooks";
import { transcode } from "buffer";
import { selectCommunityData } from "../../../features/subreddit/subredditSlice";

interface Props {
  data: DocumentData | undefined;
}

const Post: React.FC<Props> = (props) => {
  let { postId } = useParams();
  if(postId === undefined)
    postId = props?.data?.id 

  const [postData, setPostData] = useState<DocumentData | undefined>();
  const [userPostVote, setUserPostVote] = useState<DocumentData | undefined>();

  const isLoggedIn = useAppSelector(selectAuthStatus);
  const dispatch = useDispatch();
  const location = useLocation();
  const { id: subredditId } = useAppSelector(selectCommunityData);

  useEffect(() => {
    async function fetchPost() {
      const postDocRef = collection(db, "posts");

      const q = query(postDocRef, where("id", "==", postId));

      const postDoc = await getDocs(q);

      // setPostData(postDoc?.docs[0]?.data());
      onSnapshot(q, (snapshot: QuerySnapshot) => {
        console.log(snapshot.docs);
        setPostData({ ...snapshot.docs[0].data(), id: snapshot.docs[0].id });
      });
    }

    props.data ?? fetchPost();
  }, [postId, props.data]);

  const onVote = async (
    vote: number,
    event: React.MouseEvent<HTMLImageElement, MouseEvent>
  ) => {
    event.preventDefault();
    if (!isLoggedIn) {
      dispatch(toggleSignInModal());
      return;
    }

    try {
      await runTransaction(db, async (transaction) => {
        const userPostVotesRef = doc(
          db,
          "users",
          `${getUserId()}/postVotes/${postId ?? props.data?.id}`
        );

        const postRef = doc(db, "posts", postId ?? props.data?.id);

        const post = await transaction.get(postRef);

        const userPostVotes = await transaction.get(userPostVotesRef);
        let voteChange = vote;

        if (!userPostVotes.exists()) {
          const newVote = {
            id: userPostVotesRef.id,
            postId: postId ?? props.data?.id,
            subredditId,
            voteValue: vote,
          };

          transaction.set(userPostVotesRef, newVote);
        } else {
          if (userPostVotes.data().voteValue === vote) {
            voteChange *= -1;
            transaction.update(postRef, {
              voteStatus: post?.data()?.voteStatus - vote,
            });
            transaction.delete(userPostVotesRef);
          } else {
            voteChange = 2 * vote;
            transaction.update(postRef, {
              voteStatus: post?.data()?.voteStatus + 2 * vote,
            });
            transaction.update(userPostVotesRef, {
              voteValue: vote,
            });
          }
        }
        transaction.update(postRef, {
          voteStatus: post?.data()?.voteStatus + voteChange,
        });
      });
    } catch (error) {
      alert(`Could not vote on post. Please try again: ${error}`);
    }
  };

  return (
    <div
      styleName={postId ? "post-excerpt" : "post-excerpt-hover"}
      data-testid="post"
    >
      <Votes
        voteStatus={props.data?.voteStatus ?? postData?.voteStatus}
        onVote={(
          vote: number,
          event: React.MouseEvent<HTMLImageElement, MouseEvent>
        ) => onVote(vote, event)}
        postId={postId ?? props.data?.id}
      />
      <div styleName="post-excerpt__content">
        <PostAuthor
          subredditName={props.data?.subredditName ?? postData?.subredditName}
          userName={props.data?.userName ?? postData?.userName}
          createdAt={props.data?.createdAt ?? postData?.createdAt}
        />
        <div styleName="post-excerpt__container">
          <h3 styleName="post-excerpt__title">
            {props.data?.title ?? postData?.title}
          </h3>
          <p
            styleName={`post-excerpt__description ${
              !postId && "post-excerpt__overflow"
            }`}
          >
            {props.data?.description ?? postData?.description}
          </p>
          <p styleName="post-excerpt__description">
            {props.data?.link ?? postData?.link}
          </p>
          {props.data?.imageURL ||
            (postData?.imageURL && (
              <div styleName="post-excerpt__image-container">
                <img
                  styleName="post-excerpt__image"
                  src={postData?.imageURL ?? props.data?.imageURL}
                  alt="post"
                />
              </div>
            ))}
        </div>
        <div styleName="post__buttons">
          <div styleName="post-excerpt__divider"></div>
          <PostInteractions
            commentsQuantity={
              props.data?.commentsQuantity ?? postData?.commentsQuantity
            }
            postId={postId}
          />
        </div>
      </div>
    </div>
  );
};

export default CSSModules(Post, styles, {
  allowMultiple: true,
  handleNotFoundStyleName: "ignore",
});
