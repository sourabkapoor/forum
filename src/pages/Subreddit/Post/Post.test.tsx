import React from "react";
import {
  queryByTestId,
  queryByText,
  render,
  screen,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import each from "jest-each";
import Post from "./Post"

describe("Post", () => {
  const mockProps = {
    data: {
      title: "mock title",
      description: "mock description",
      voteStatus: 0,
      userVoteValue: 0
    },
    commentsQuantity: 0,
  };

  // test("snapshot", () => {
  //   const { asFragment } = render(<Post {...mockProps} />);

  //   expect(asFragment()).toMatchSnapshot();
  // });
  // test("post rendering correctly", () => {
  //   render(<Post {...mockProps} />);

  //   const postTitle = screen.queryByText("mock title");
  //   const postDescription = screen.queryByText("mock description");
  //   const postVoteStatus = screen.queryByText("0");

  //   expect(postTitle).toBeInTheDocument();
  //   expect(postDescription).toBeInTheDocument();
  //   expect(postVoteStatus).toBeInTheDocument();
  // });

  // test("post having hover border", () => {
  //   render(<Post {...mockProps} />);

  //   const post = screen.queryByTestId("post");
  //   expect(post?.classList.contains("post-excerpt-hover")).toBe(true);
  // });
  // test("Redirecting user to the post page when clicked", 9) => {
  //   return (
  //     <MemoryRouter>

  //     </MemoryRouter>
  //   )
  // }
});
