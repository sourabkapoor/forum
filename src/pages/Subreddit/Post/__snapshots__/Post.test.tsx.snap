// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post snapshot 1`] = `
<DocumentFragment>
  <div
    class="post-excerpt"
    data-testid="post"
  >
    <div
      class="votes"
    >
      <div
        class="votes__vote votes__vote_type_upvote"
      >
        <svg
          class="votes__icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.781 2.375c-.381-.475-1.181-.475-1.562 0l-8 10A1.001 1.001 0 0 0 4 14h4v7a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1v-7h4a1.001 1.001 0 0 0 .781-1.625l-8-10zM15 12h-1v8h-4v-8H6.081L12 4.601 17.919 12H15z"
          />
        </svg>
      </div>
      <p
        class="votes__likes"
      >
        0
      </p>
      <div
        class="votes__vote votes__vote_type_downvote"
      >
        <svg
          class="votes__icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20.901 10.566A1.001 1.001 0 0 0 20 10h-4V3a1 1 0 0 0-1-1H9a1 1 0 0 0-1 1v7H4a1.001 1.001 0 0 0-.781 1.625l8 10a1 1 0 0 0 1.562 0l8-10c.24-.301.286-.712.12-1.059zM12 19.399 6.081 12H10V4h4v8h3.919L12 19.399z"
          />
        </svg>
      </div>
    </div>
    <div
      class="post-excerpt__content"
    >
      <div
        class="post-excerpt__info"
      >
        <span
          class="post-excerpt__posted-by"
        >
          Posted by
        </span>
        <p
          class="post-excerpt__author"
        >
          /u/TheWatchingBug
        </p>
        <span
          class="post-excerpt__date"
        >
          4 months ago
        </span>
      </div>
      <div
        class="post-excerpt__container"
      >
        <h3
          class="post-excerpt__title"
        >
          mock title
        </h3>
        <p
          class="post-excerpt__description"
        >
          mock description
        </p>
      </div>
      <div
        class="post__buttons"
      >
        <div
          class="post-excerpt__divider"
        />
        <div
          class="post-excerpt__interactions"
        >
          <div
            class="post-excerpt__interaction"
          >
            <svg
              class="post-excerpt__icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 2H4c-1.103 0-2 .897-2 2v12c0 1.103.897 2 2 2h3v3.767L13.277 18H20c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zm0 14h-7.277L9 18.233V16H4V4h16v12z"
              />
            </svg>
            <span
              class="post-excerpt__action"
            >
              0 Comments
            </span>
          </div>
          <div
            class="post-excerpt__interaction"
          >
            <svg
              class="post-excerpt__icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M448 248L288 96v85.334C138.666 202.667 85.333 309.334 64 416c53.333-74.666 117.333-108.802 224-108.802v87.469L448 248z"
              />
            </svg>
            <span
              class="post-excerpt__action"
            >
              Share
            </span>
          </div>
          <div
            class="post-excerpt__interaction"
          >
            <svg
              class="post-excerpt__icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M336 0H48C21.49 0 0 21.49 0 48v464l192-112 192 112V48c0-26.51-21.49-48-48-48zm0 428.43l-144-84-144 84V54a6 6 0 0 1 6-6h276c3.314 0 6 2.683 6 5.996V428.43z"
              />
            </svg>
            <span
              class="post-excerpt__action"
            >
              Save
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
